sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ControlFlowException;
+import com.oracle.truffle.api.nodes.RootNode;
+import com.oracle.truffle.llvm.nodes.intrinsics.llvm.LLVMBuiltin;
+import com.oracle.truffle.llvm.runtime.LLVMFunctionDescriptor;
+import com.oracle.truffle.llvm.runtime.LLVMLanguage;
+import com.oracle.truffle.llvm.runtime.interop.access.LLVMInteropType;
+import com.oracle.truffle.llvm.runtime.memory.LLVMStack;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMLoadNode;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMStoreNode;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMNativePointer;
+import com.oracle.truffle.llvm.runtime.pointer.LLVMPointer;
+import com.oracle.truffle.llvm.runtime.types.FunctionType;
+import com.oracle.truffle.llvm.runtime.types.PointerType;
+import com.oracle.truffle.llvm.runtime.types.Type;
+
+public class LLVMPThreadIntrinsics {
+    public static PrintWriter debugOut = null;
+
+    @CompilerDirectives.TruffleBoundary
+    private static void printDebug(String str) {
+        // create debug output tool
+        if (debugOut == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            try {
+                debugOut = new PrintWriter(new FileWriter("/home/florian/debug out"));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        debugOut.println(str);
+        debugOut.flush();
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadCreate extends LLVMBuiltin {
+        @Child
+        LLVMStoreNode store = null;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object thread, Object attr, Object startRoutine, Object arg) {
+            // create store node
+            if (store == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                store = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createStoreNode(LLVMInteropType.ValueKind.I64);
+            }
+
+            // create thread for execution of function
+            Thread t = LLVMLanguage.getLLVMContextReference().get().getEnv().createThread(() -> {
+                CompilerDirectives.transferToInterpreter();
+                // RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(getLLVMLanguage()));
+                RootCallTarget callTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(LLVMLanguage.getLanguage()));
```

Comment:
Building a new `RootCallTarget` each time `pthread_create` is called is not good for performance. Since the function to call is passed as an argument to the `callTarget` in any case, you can create `callTarget` only once, store it in the context, and then re-use it wherever `pthread_create` is called. Then you also do not need the `transferToInterpreter` here.